File: day-2/rust/src/solution/model/dimension.rs
use std::arch::aarch64::vorn_u8;
use std::cmp;

pub struct Dimension {
  pub length: usize,
  pub width: usize,
  pub height: usize,
}

const DIMENSION_DELIMITER: char = 'x';

impl Dimension {
  pub fn new (length: usize, width: usize, height: usize) -> Dimension {
    Dimension { length, width, height }
  }

  pub fn from_string(s: &str) -> Dimension {
    let dims = s.split(DIMENSION_DELIMITER).collect::<Vec<&str>>();

    let l = *dims.get(0).expect(&format!("Dimension {} missing length", s));
    let w = *dims.get(1).expect(&format!("Dimension {} missing width", s));
    let h = *dims.get(2).expect(&format!("Dimension {} missing height", s));

    let l_num = l.parse::<usize>().expect(&format!("Invalid length {}", l));
    let w_num = w.parse::<usize>().expect(&format!("Invalid width {}", w));
    let h_num = h.parse::<usize>().expect(&format!("Invalid height {}", h));

    Dimension::new(l_num, w_num, h_num)
  }

  pub fn total_wrapping_paper(self) -> usize {
    let mut total_surface_area = 0;
    let smallest_surface_area = cmp::min(
      self.length * self.width,
      cmp::min(
        self.width * self.height,
        self.height * self.length,
      ),
    );
    total_surface_area += (2 * self.length * self.width);
    total_surface_area += (2 * self.height * self.width);
    total_surface_area += (2 * self.length * self.height);
    total_surface_area += smallest_surface_area;
    total_surface_area
  }

  pub fn total_ribbon(self) -> usize {
    let smallest_perimeter = cmp::min(
      2 * self.length + 2 * self.width,
      cmp::min(
        2 * self.width + 2 * self.height,
        2 * self.height + 2 * self.length,
      )
    );
    let volume = self.width * self.height * self.length;
    let total_ribbon = volume + smallest_perimeter;
    total_ribbon
  }
}

File: day-2/go/model/dimension.go
package dimension

import (
	"fmt"
	"strconv"
	"strings"
)

type Dimension struct {
	Length int64
	Width  int64
	Height int64
}

const DIMENSION_DELIMITER = "x"

func FromString(s string) (Dimension, error) {
	dims := strings.Split(s, DIMENSION_DELIMITER)
	length, err := strconv.ParseInt(dims[0], 10, 64)
	if err != nil {
		fmt.Println("Error converting string to length:", err)
		return Dimension{}, err
	}
	width, err := strconv.ParseInt(dims[1], 10, 64)
	if err != nil {
		fmt.Println("Error converting string to width:", err)
		return Dimension{}, err
	}
	height, err := strconv.ParseInt(dims[2], 10, 64)
	if err != nil {
		fmt.Println("Error converting string to height:", err)
		return Dimension{}, err
	}
	return Dimension{Length: length, Width: width, Height: height}, nil
}

func (d Dimension) SurfaceArea() int64 {
	total := int64(0)
	total += 2 * (d.Length * d.Width)
	total += 2 * (d.Width * d.Height)
	total += 2 * (d.Height * d.Length)
	return total
}

func (d Dimension) SmallestSideArea() int64 {
	return min(
		(d.Length * d.Width),
		(d.Width * d.Height),
		(d.Height * d.Length),
	)
}

func (d Dimension) SmallestPerimeter() int64 {
	return min(
		(2*d.Length + 2*d.Width),
		(2*d.Width + 2*d.Height),
		(2*d.Height + 2*d.Length),
	)
}

func (d Dimension) Volume() int64 {
	return d.Length * d.Width * d.Height
}
